extensions/2.0/Vendor/FB_material_variants/README.md

```diff
@@ -0,0 +1,98 @@
+# FB_material_variants
+
+## Contributors
+
+- Pär Winzell, Facebook, [@zellski](https://twitter.com/zellski)
+- Renee Rashid, Facebook, [@debuggrl](https://github.com/debuggrl)
+
+## Status
+
+Draft
+
+## Dependencies
+
+Written against the glTF 2.0 spec.
+
+## Overview
+
+This extension allows for a compact glTF representation of multiple material variants of an asset, structured to allow low-latency switching at runtime.
+
+A typical use case is digital commerce, where a user might be presented with e.g. a pair of sneakers and the ability to switch between different colours.
+
+## Tagged Variants
+
+We introduce a simple tag-based extension scheme, that allows for high-level runtime swapping between which `material` is used to shade a given `mesh primitive`: the extension root contains a mandatory `mapping` property, which is an array of objects, each one associating some set of tags with a material reference.
+
+Imagine a sneaker with shoelace holes that are made from materials that depend on the overall shoe colour in non-obvious ways:
+
+| Tags                               | Material                        |
+| ---------------------------------- | ------------------------------- |
+| `sneaker_yellow`, `sneaker_orange` | `shoelace_hole_material_brown`  |
+| `sneaker_red`                      | `shoelace_hole_material_purple` |
+| `sneaker_black`                    | `shoelace_hole_material_yellow` |
+
+(_the authors of this spec are not product designer, apologies for the dubious colour choices_)
```

Comment:
Would you consider exposing in metadata all the choices that the users can make and all the values that they can specify? There might be more than just a single set of variants.
Those could drive the UI. And even if the UI is defined by the database this could be a way to ensure that the model is capable of representing all the same options exposed to the user.

If this ends up being applied, I would say that the tagnames would benefit from a well-defined format that combines both the property name together with the selected value. For example underscores might not be allowed for property  and value ids and be reserved only for concatenating those together.

Reply:
> Would you consider exposing in metadata all the choices that the users can make and all the values that they can specify?

Yea I think we would need to expose multiple tag 'sets' through metadata. When we talk about driving UI with content from the glTF extension I get a bit wary as we should try to keep the data stored in the extension as business-logic free as possible.

> If this ends up being applied, I would say that the tagnames would benefit from a well-defined format that combines both the property name together with the selected value.

I think having a clear and standardized way of tagging will make it easier for viewers to implement the spec so I can see this being useful regardless.

Reply:
This has been brought up several times and consensus so far is to keep business-logic out of this extension. It can definitely lead to a rabbit hole of complexity as soon as there are thousands of combinations or rules start being needed (e.g The red laces can go with the yellow or red sole, but not the blue sole unless the purple heel is chosen).

It also puts a big burden on the 3D authoring tools. We can't expect 3D modelling tools to support defining the metadata of all possible configurations.